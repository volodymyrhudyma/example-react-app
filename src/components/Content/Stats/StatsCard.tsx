import React, { Component, SyntheticEvent } from 'react';
import {
  CardHeading,
  CardBody,
  ActionsWrapper,
  Count,
  Period,
} from './styled/StatsCard.styled';
import icons from 'src/constants/icons';
import Title from 'src/ui-kit/Title';
import Card from 'src/ui-kit/Card';
import Icon from 'src/ui-kit/Icon';
import Dropdown from 'src/ui-kit/Dropdown';
import { IArticle } from 'src/types';
import { withTranslation, WithTranslation } from 'react-i18next';

interface IStatsCardProps extends WithTranslation {
  title: string;
  count: number;
  period: string;
  type: string;
  articles: IArticle[];
  readArticle: (article: IArticle) => void;
}

interface IStatsCardState {
  dropdownShown: boolean;
}

class StatsCard extends Component<IStatsCardProps, IStatsCardState> {
  constructor(props: IStatsCardProps) {
    super(props);
    this.state = {
      dropdownShown: false,
    };
  }
  toggleDropdown = (e?: SyntheticEvent) => {
    if (e) {
      e.preventDefault();
    }
    this.setState(({ dropdownShown }) => ({
      dropdownShown: !dropdownShown,
    }));
  };
  onReadAllClick = async () => {
    const { articles, readArticle } = this.props;
    for (const article of articles) {
      await readArticle(article);
    }
    console.log('toggle');
    this.toggleDropdown();
  };
  render() {
    const { t, title, count, period, type } = this.props;
    const { dropdownShown } = this.state;
    return (
      <Card width="300px">
        <CardHeading>
          <Title type="h6">{title}</Title>
          <ActionsWrapper href="#" onClick={this.toggleDropdown}>
            <Icon src={icons.more} />
          </ActionsWrapper>
          {dropdownShown && (
            <Dropdown
              close={this.toggleDropdown}
              action={this.onReadAllClick}
              buttonText={t('articles.readAll')}
            />
          )}
        </CardHeading>
        <CardBody>
          <Count type={type}>{count}</Count>
          <Period>{period}</Period>
        </CardBody>
      </Card>
    );
  }
}

export default withTranslation()(StatsCard);
