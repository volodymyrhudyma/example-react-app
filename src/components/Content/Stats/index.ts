import Component from './Stats';
import { connect } from 'react-redux';
import { IAppState, IArticle } from 'src/types';
import { readArticle } from 'src/actions/article';
import { withTranslation } from 'react-i18next';

const mapStateToProps = (state: IAppState) => ({
  articles: state.article.articles.data,
});

const mapDispatchToProps = (dispatch: any) => {
  return {
    readArticle: (article: IArticle) => dispatch(readArticle(article)),
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withTranslation()(Component));
