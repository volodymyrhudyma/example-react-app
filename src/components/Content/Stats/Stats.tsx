import React, { Component } from 'react';
import { Wrapper, CardWrapper } from './styled/Stats.styled';
import StatsCard from './StatsCard';
import { IArticle } from 'src/types';
import { WithTranslation } from 'react-i18next';

interface IStatsProps extends WithTranslation {
  articles: IArticle[];
  readArticle: (article: IArticle) => void;
}

class Stats extends Component<IStatsProps> {
  render() {
    const { t, articles, readArticle } = this.props;
    return (
      <Wrapper>
        <CardWrapper>
          <StatsCard
            title={t('articles.total')}
            count={articles.length}
            period={t('articles.totalPeriod')}
            type="overall"
            articles={articles}
            readArticle={readArticle}
          />
        </CardWrapper>
        <StatsCard
          title={t('articles.unread')}
          count={articles.filter(article => !article.isRead).length}
          period={t('articles.unreadPeriod')}
          type="unread"
          articles={articles}
          readArticle={readArticle}
        />
      </Wrapper>
    );
  }
}

export default Stats;
