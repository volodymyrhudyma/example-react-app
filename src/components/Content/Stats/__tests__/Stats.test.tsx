import { shallow } from 'enzyme';
import React from 'react';
import Stats from '../Stats';
import i18n from 'i18next';

describe('Stats with Enzyme and Jest', () => {
  it('should render correctly', () => {
    const component = shallow(
      <Stats
        t={jest.fn()}
        i18n={i18n}
        tReady={false}
        articles={[]}
        readArticle={jest.fn()}
      />
    );
    expect(component).toMatchSnapshot();
  });
});
