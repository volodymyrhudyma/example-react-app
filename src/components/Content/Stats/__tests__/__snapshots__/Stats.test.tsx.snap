// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stats with Enzyme and Jest should render correctly 1`] = `
<styled.div>
  <styled.div>
    <withI18nextTranslation(StatsCard)
      articles={Array []}
      count={0}
      readArticle={[MockFunction]}
      type="overall"
    />
  </styled.div>
  <withI18nextTranslation(StatsCard)
    articles={Array []}
    count={0}
    readArticle={[MockFunction]}
    type="unread"
  />
</styled.div>
`;
